Description: Highlight current day.
Author: Michael Meskes <meskes@debian.org>

diff -rNu bsdmainutils/usr.bin/ncal/ncal.c bsdmainutils/usr.bin/ncal/ncal.c
--- bsdmainutils/usr.bin/ncal/ncal.c	2009-11-22 18:52:17.000000000 +0100
+++ bsdmainutils/usr.bin/ncal/ncal.c	2009-11-22 18:51:18.000000000 +0100
@@ -400,7 +400,7 @@
 		printeaster(y, flag_julian_cal, flag_orthodox);
 	else if (argc == 1 || flag_hole_year) {
 		/* disable the highlight for now */
-		today = 0;
+		/* today = 0; */
 		if (flag_backward)
 			printyearb(y, flag_julian_day);
 		else
@@ -820,12 +820,13 @@
 	for (i = 0; i != 7; i++) {
 		l = high = 0;
 		for (j = firstm + i, k = 0; j < last; j += 7, k += dw) {
-			if (j == today && (term_so != NULL && term_se != NULL)) {
-				l = strlen(term_so);
+			if (j >= first) {
 				if (jd_flag)
 					dt.d = j - jan1 + 1;
 				else
-					sdateb(j, &dt);
+					sdate(j, &dt);
+			if (j == today && (term_so != NULL && term_se != NULL)) {
+				l = strlen(term_so);
 				/* separator */
 				mlines->lines[i][k] = ' ';
 				/* the actual text */
@@ -839,11 +840,6 @@
 				l = high = strlen(term_se) + strlen(term_so);
 				continue;
 			}
-			if (j >= first) {
-				if (jd_flag)
-					dt.d = j - jan1 + 1;
-				else
-					sdate(j, &dt);
 				memcpy(mlines->lines[i] + k + l,
 				       ds + dt.d * dw, dw);
 			} else
@@ -951,12 +947,13 @@
 		l = high = 0;
 		for (j = firsts + 7 * i, k = 0; j < last && k != dw * 7;
 		    j++, k += dw) { 
-			if (j == today && (term_so != NULL && term_se != NULL)) {
-				l = strlen(term_so);
+			if (j >= first) {
 				if (jd_flag)
 					dt.d = j - jan1 + 1;
 				else
 					sdateb(j, &dt);
+			if (j == today && (term_so != NULL && term_se != NULL)) {
+				l = strlen(term_so);
 				/* separator */
 				mlines->lines[i][k] = ' ';
 				/* the actual text */
@@ -970,11 +967,6 @@
 				l = high = strlen(term_se) + strlen(term_so);
 				continue;
 			}
-			if (j >= first) {
-				if (jd_flag)
-					dt.d = j - jan1 + 1;
-				else
-					sdateb(j, &dt);
 				memcpy(mlines->lines[i] + k + l,
 				       ds + dt.d * dw, dw);
 			} else
